Update SDK for chrome extension demo

knewkirk opened the pull request:
## Highlights:
- more options for custom messaging implementation (no SW)
- better promise handling / async logic for requests made before login
- basic `on` implementation to allow for login waiting
- simplified redirect page logic, no more custom route needed, can now redirecting to primary page location

_Sorry, I know it's getting a bit confusing/unruly - splitting up logic soon. Main changes are just in [wally-connector.ts](https://github.com/Wally-xyz/wallet-connector/compare/redirect-messaging?expand=1#diff-b178b1dbd26bbc533e1e1d249e660feec968f124ba9b62eaabb14d200362fae7)_

https://user-images.githubusercontent.com/2947034/203455298-c9d73911-8b2c-4c1c-b012-7b70eec31c69.mov

[Extension updates here](https://github.com/Wally-xyz/extension-tester/pull/1)

wchan2 commented on src/types.ts:
```diff
@@ -17,22 +17,64 @@ export type SignedMessage = {
+   * tab when the secondary
+   * @default false
+   */
+  disableSharedWorker?: boolean;
```
Q: Just curious, in what scenario do we need this?

knewkirk replied:
Mostly just for the chrome extension - if there's a situation where the shared worker can't be used or isn't working, I thought it would be nice to just disable instead of spewing console errors

wchan2 replied:
I see and disabling won't break any intentional behaviors for the SDK?

knewkirk replied:
the SDK _works_ without it - the shared worker just informs the primary window when the secondary redirect has finished. the user can just refresh the main window and everything is fine

again, just for the chrome extension - i'll add a note

wchan2 replied:
Sounds good. This is non-blocking for me. Just want to understand it a little deeper. Don't know what has been done for the chrome extension overall.

wchan2 commented on src/types.ts:
```diff
@@ -17,22 +17,64 @@ export type SignedMessage = {
+   * is set as a regex.
+   * @default false
+   */
+  redirectToCurrentLocation?: boolean;
```
RegEx can also be an exact match too right? In that scenario, it's a little redundant to have `redirectToCurrentLocation` too when they can set it to the `window.location.href`.

knewkirk replied:
yeah I agree for sure - was I was thinking, was do you think there's a case though where they could initialize wally with something like the home page of the app, but then the user navigates around to another page where they login from, where they would rather have the user redirected to?

wchan2 replied:
Actually, I'm not sure there's a use case and I think Mayank and I are on slightly different sides of the spectrum. As an engineer, I think all the flexibility makes it more confusing for me to integrate the SDK but if there's a single well understood way of why it's necessary, then I can implement the said custom logic myself. The flexibility takes away any clear direction of whats the simplest and best approach.

knewkirk replied:
yeah for sure, i'll take it out. i think the tough part is that there isn't really _any_ use case yet because no one's using it and we aren't sure what sorts of flexibility will be necessary

wchan2 commented on src/wally-connector.ts:
```diff
@@ -17,54 +16,100 @@ import {
+        cb(value);
       } else {
-        if (verbose) {
-          console.log('Setting auth token');
-        }
-        this.setAuthToken(token);
+        cb();
```
If value is optional or `undefined`, you can just simplify with a single `cb(value)`.

knewkirk replied:
the important part is that we _don't_ want to pass anything when emitting messages other than `accountsChanged` - I'm not entirely sure which messages an application would use, and there isn't really a spec, so I'd rather err on the side of no value rather than an incorrect or misinterpreted value

wchan2 commented on src/wally-connector.ts:
```diff
@@ -183,6 +246,8 @@ class WallyConnector {
         const data = await resp.json();
         this.setAuthToken(data.token);
+        this.selectedAddress = data.wallet;
+        this.onTokenFetched && this.onTokenFetched(data.wallet);
```
Q: Does the following work 🤔 ?

```suggestion
        this.onTokenFetched?(data.wallet);
```

knewkirk replied:
unfortunately not ☹️ 

![Screen Shot 2022-11-28 at 11 58 55 AM](https://user-images.githubusercontent.com/2947034/204369491-719051cb-720d-46b8-ab2d-e76d49f395e0.png)

wchan2 commented on src/wally-connector.ts:
```diff
@@ -216,7 +281,7 @@ class WallyConnector {
 
-    if (closeWindow) {
+    if (!this.disableRedirectClose) {
       window.setTimeout(window.close, 1000);
```
Q: What do we typically use the timeout for?

knewkirk replied:
🤷  minor ux - it's a little bit less jarring imo than just closing immediately, and there might be some last bits of cleanup or functions that need to finish before it closes

thoughts?

wchan2 replied:
Nothing. It's just me; whenever I see a timeout, it could potentially be a race and that's why people use it. So I'm naturally just picking that up.

wchan2 commented on src/wally-connector.ts:
```diff
@@ -267,6 +332,13 @@ class WallyConnector {
+  /**
+   * @deprecated - see this.request()
+   */
+  public sendAsync(req: any): Promise<any> {
+    return this.request(req);
+  }
```
Q: Does the following not work?

```suggestion
  public async sendAsync(req: any) {
    return this.request(req);
  }
```

knewkirk replied:
👍

wchan2 commented on src/wally-connector.ts:
```diff
@@ -325,6 +397,33 @@ class WallyConnector {
+  ): Promise<MethodResponse<T> | null> {
+    return new Promise((resolve, reject) => {
+      if (!this.disableLoginOnRequest && !this.isLoggingIn) {
+        this.login().then(() => {
+          resolve(this.request(req));
+        });
```
Q: Why not use `async` and `await`? Seeing a lot of code that constructs promises.

knewkirk replied:
left a comment above - to me it's more explicit than relying on `await` + `return`, which in my experience has sometimes not worked out the way i intended. i _think_ there might also still be some iffy support.

what's the advantage of using async/await?

wchan2 replied:
Either or is fine. Most times the async and await will come out cleaner. async-await and promises are pretty much identical except one is more procedural and arguably a little more "OO-ish". In order to achieve the above, you'd put the function together with await and then after that do the emitting if they're supposed to be done together so something like the following. 

```javascript
async function myFunc() {
  await login();
  emit(...);
}

// in the main caller
await myFunc();
```

knewkirk replied:
oh yeah i mean i'm definitely familiar with how async/await works, i was just commenting on how it gets a little bit more complicated when you're resolving promises inside of other callbacks - and I don't think it maps very cleanly:
```js
return new Promise((resolve, reject) => {
  const listener = () => {
    this.removeListener(EmitterMessage.ACCOUNTS_CHANGED, listener);
    resolve(this.request(req));
  };
  this.on(EmitterMessage.ACCOUNTS_CHANGED, listener);
})
```
tbh i don't know if i'm comfortable refactoring this logic just for the sake of being a little cleaner

wchan2 replied:
While a promise can have the behavior above, it's actually a little hacky because say if you `await` the promise that is returned but the callback is never triggered, then it would block. It prevents flexibility in the code, because then a promise always needs to be used for every single scenario and if it doesn't resolve, the chaining of the promises would never resolve. 

What you want instead is probably a callback function to passed into the function and decorate that function with the remove listener and pass the return value into that same callback.

wchan2 commented on src/wally-connector.ts:
```diff
@@ -436,7 +535,7 @@ class WallyConnector {
     }
 
-    return null;
+    return Promise.reject(`Invalid response for ${method}`);
```
I think reject usually passes in an error.
```suggestion
    return Promise.reject(new Error()`Invalid response for ${method}`);
```

knewkirk replied:
👍

wchan2 commented on src/wally-connector.ts:
```diff
@@ -482,7 +581,7 @@ class WallyConnector {
       );
     }
-    return null;
+    return Promise.reject(`Invalid response for ${method}`);
```
Ditto here:

```suggestion
    return Promise.reject(new Error(`Invalid response for ${method}`));
```

wchan2 reviewed (commented):
Mostly minor comments.

knewkirk commented on src/index.ts:
```diff
@@ -1,36 +1,56 @@
 }
 
+const checkInjected = () => {
+  if (!window.wally) {
```
so do something like `window.WallyConnector.wally`? wouldn't that be pretty much the same thing? I know it's not perfect, but I don't think one global is _too_ bad.... we're going to need some sort of global to access the instance throughout the consuming application, and it's how other providers work, i.e. `window.ethereum` for metamask and gamestop.

wchan2 replied:
Yea, I think it's relatively minor but having too many globals can affect a developer's runtime so that's why it could be nice to namespace. `WallyConnector` is actually already global and so using the class name as the namespace could be a good place. I think the thing I'd caution on is that we should to be aware of what would interrupt the developer's environment and not strictly just ours since we're just building the SDK. They own the environment.

knewkirk replied:
was looking through the code again and we don't need the global anymore 🤷

knewkirk commented on src/index.ts:
```diff
@@ -1,36 +1,56 @@
 declare global {
   // eslint-disable-next-line no-var
-  var wally: WallyConnector
+  var wally: WallyConnector;
```
typescript global declarations only work with `var` 🤷 
![Screen Shot 2022-11-28 at 11 08 10 AM](https://user-images.githubusercontent.com/2947034/204360586-0a2b99fc-65bb-445c-83ae-d2344de34c57.png)

wchan2 replied:
That's interesting 😄 .

knewkirk commented on src/index.ts:
```diff
@@ -1,36 +1,56 @@
-}
+};
+
+export const login = (): Promise<void> => {
```
sure 👍 

I know you've got a few comments below about promises - any particular reason you aren't a fan? For simple stuff like this, I totally get it. But in some cases below I feel a little bit better explicitly using `resolve` and `reject` instead of an implicit `return`, which might not end up doing what you intend them to depending on the context

knewkirk commented:
Thanks for the review! I think I addressed most of the feedback - still a couple open questions around promises, options, minor details - I responded in threads, lmk if anything is unclear, happy to discuss further/make some more changes

Randy1Burrell reviewed (approved):
I see no problems

wchan2 reviewed (approved):
Looks good to me; most of the PR comments are minor and non-blocking with the exception that I'd recommend taking a deeper look into. Mainly [this comment](https://github.com/Wally-xyz/wallet-connector/pull/25#discussion_r1030007300). It does affect actual behavior of whats happening and potentially even whats intentional.
